django/contrib/contenttypes/models.py

```diff
@@ -64,24 +65,29 @@ def get_for_models(self, *models, for_concrete_models=True):
         Given *models, return a dictionary mapping {model: content_type}.
         """
         results = {}
-        # Models that aren't already in the cache.
-        needed_app_labels = set()
-        needed_models = set()
+        # Models that aren't already in the cache grouped by app labels.
+        needed_models = defaultdict(set)
         # Mapping of opts to the list of models requiring it.
         needed_opts = defaultdict(list)
         for model in models:
             opts = self._get_opts(model, for_concrete_models)
             try:
                 ct = self._get_from_cache(opts)
             except KeyError:
-                needed_app_labels.add(opts.app_label)
-                needed_models.add(opts.model_name)
+                needed_models[opts.app_label].add(opts.model_name)
                 needed_opts[opts].append(model)
             else:
                 results[model] = ct
         if needed_opts:
             # Lookup required content types from the DB.
-            cts = self.filter(app_label__in=needed_app_labels, model__in=needed_models)
+            condition = Q(
+                *(
+                    Q(("app_label", app_label), ("model__in", models), _connector=Q.AND)
+                    for app_label, models in needed_models.items()
+                ),
+                _connector=Q.OR,
+            )
```

Comment:
Minor but this could have likely be simplified by using `reduce` to avoid the private `_connector` usage

```python
condition = reduce(
    (Q(app_label=app_label, model__in=models) for app_label, models in needed_models)
, operator.or_)
```


In all cases `Q(("app_label", app_label), ("model__in", models), _connector=Q.AND)` can be simplified to `Q(app_label=app_label, model__in=models)` since `_connector` defaults to `Q.AND`.

Reply:
@charettes Thanks :+1: I removed unnecessary connector, see #15511. As far as I'm aware we now prefer non-kwargs constructions for internal usage, see 9662193aea2ee982bc8e553c62499aca5e606755 and #14699,